#570 Implement default env creation

kangsuhyun-yanolja opened the pull request:
**This PR doesn't yet fully implement the requirements of issue #570. 
The Docker process and Endpoint display still need to be modified.**

It includes the following changes:
- Add three default environments (development, staging, production) when a new app is created.
- Introduce an Environment collection.
- Add a 'Publish' button to the Playground page.
- Introduce a Publish Variant modal.
- Update the Environment collection when publishing is triggered.
<img width="1174" alt="스크린샷 2023-09-15 오후 3 07 47" src="https://github.com/Agenta-AI/agenta/assets/124246127/a5301dc0-0392-447c-b639-60bbab2c07d1">

mmabrouk commented:
Merged main into this due to it containing a hot fix.

mmabrouk reviewed (changes requested):
@kangsuhyun-yanolja, thank you for your PR!

I've reviewed the changes and added some inline comments. Here's a summary:

1. **Publishing Unsaved Variants**:
    - **Issue**: Variants can be published without being saved to the database.
    - **Reproduction**:
        1. Create a variant with the '+' button.
        2. Hit publish.
       Outcome: The variant is published but not stored in the database.
    - **Solution**:
        - Frontend: Disable the publish button until the variant is saved.
        - Backend: Validate the variant's existence before deployment.

2. **Overwriting Warning**:
    - **Issue**: No warning is provided when overwriting an environment.
    - **Solution**: Display a warning in the publishing modal if an environment is about to be overwritten.

3. **Environment Reset**:
    - **Issue**: Environments aren't reset upon variant deletion.
    
4. **Environment Cleanup**:
    - **Issue**: Deleting an app doesn't remove its associated environments.

**Next Steps**:

1. We need to implement an endpoint to fetch the variant parameters using app and environment names.
2. We need to change the endpoint view to list environments (instad of variants). Display the code of the deployed variant for each environment.
3. Add a section in the endpoint view guiding the use of environment parameters via the new endpoint.

kangsuhyun-yanolja commented:
@mmabrouk Thank you for the thorough review. I have addressed the issues you pointed out.

yeokyeong-yanolja commented on agenta-backend/agenta_backend/services/db_manager.py:
```diff
@@ -89,6 +88,18 @@ async def add_variant_based_on_image(
         raise ValueError("App variant with the same name already exists")
 
+    is_new_app = not await does_app_exist(app_variant.app_name, **kwargs)
+    if is_new_app:
```
I understand it's a demo version, but I believe we should add a feature that allows us to create new environments for existing apps—especially those made before this pull request—to fully support the publishing functionality.

mmabrouk replied:
One solution for this is to create the default environments not here but instead at 
`async def list_environments(app_name: str, **kwargs: dict) -> List[Environment]:`

Basically we modify it so that if the list of environments is empty we call a function `initialize_environements` which creates the environments below

kangsuhyun-yanolja replied:
Thank you for the suggestion; I'll modify the code accordingly. And thanks to Yoon for pointing it out.

mmabrouk commented:
Thanks a lot for the changes @kangsuhyun-yanolja!
```
    Publishing Unsaved Variants:
        Issue: Variants can be published without being saved to the database.
        Reproduction:
            Create a variant with the '+' button.
            Hit publish.
            Outcome: The variant is published but not stored in the database.
        Solution:
            Frontend: Disable the publish button until the variant is saved.
            Backend: Validate the variant's existence before deployment.
```

This problem is still not completely solved. Please see the video. I think the solution is to implement `Disable the publish button until the variant is saved`

https://github.com/Agenta-AI/agenta/assets/4510758/cf718190-1e7a-4922-b99c-c6dddcabd729

mmabrouk commented:
I have made a small fix to the cleanup after app deletion logic and slightly renamed some function. Hope that is alright.

mmabrouk reviewed (changes requested):
Thanks a lot for the changes @kangsuhyun-yanolja 
It think there are two missing issues:
1. The one with the disabling the publish button in the UI before saving (see note)
2. The one comment from @yeokyeong-yanolja 
After that the next steps are:
1. Implementing an endpoint to fetch the parameters for the deployed variant on an environment
2. Add a section in the endpoint view guiding the use of environment parameters via the new endpoint.

kangsuhyun-yanolja commented:
@mmabrouk Sorry I forgot about the button😂 Thank you again for the review. I've addressed these issues:

> 1. The one with the disabling the publish button in the UI before saving (see note)
> 2. The one comment from @yeokyeong-yanolja
>    After that the next steps are:

mmabrouk commented:
Thanks a lot @kangsuhyun-yanolja ! Yoon already implemented point 3 here (https://github.com/kangsuhyun-yanolja/agenta/pull/1) so the only thing that is left is adding point four and we can close this issue.
I am working right now on designing how we can solve the larger issue of deploying the variants without parameters so that the user can use the endpoint proj_name/env_name/generate with the inputs only. I will create an issue for this today.

mmabrouk commented:
I agree that hiding the button is a better UX. Great idea!

kangsuhyun-yanolja commented:
@mmabrouk I've tried to implement point 4 but not sure if I fully understand it.
I have replaced the variant with the environment on the Endpoint page like this:

![ezgif com-video-to-gif](https://github.com/Agenta-AI/agenta/assets/124246127/783b1e86-9706-4aed-8886-25f339322bac)

Please let me know if I've misunderstood something 🙂

mmabrouk reviewed (approved):
Thanks a lot for the great work @kangsuhyun-yanolja !
Optionally, we can add later a section to the endpoint view where we show how to use the endpoint for fetching the deployed prompt.

mmabrouk commented:
Closes #570 
Closes #629
